File: snippet/leetcode/balanced_tree.c
#include <stdlib.h>	/* abs */

static int
depth_of_tree(bin_tree *root)
{
	int dl, dr;

	if (root == NULL)
		return 0;
	dl = depth_of_tree(root->left);
	dr = depth_of_tree(root->right);
	if (dl == -1 || dr == -1 || abs(dl - dr) > 1)
		return -1;
	return 1 + (dl > dr) ? dl : dr;
}

int
is_tree_balanced(bin_tree *root)
{
	return depth_of_tree(root) >= 0;
}

File: snippet/leetcode/BalancedTree.java
int depthOfBalancedTree(TreeNode root)
{
    if (root == null)
	return 0;
    
    int dl = depthOfBalancedTree(root.left);
    int dr = depthOfBalancedTree(root.right);
    if (dl == -1 || dr == -1 || Math.abs(dl - dr) > 1)
            return -1;

    return 1 + (dl > dr ? dl : dr);
}

bool isTreeBalanced(TreeNode root)
{
    return depthOfBalancedTree(root) >= 0;
}
